reflex/state.py

```diff
@@ -1065,26 +1315,126 @@ def get_state(self, token: str) -> State:
             The state for the token.
         """
         if self.redis is not None:
-            redis_state = self.redis.get(token)
+            redis_state = await self.redis.get(token)
             if redis_state is None:
-                self.set_state(token, self.state())
-                return self.get_state(token)
+                await self.set_state(token, self.state())
+                return await self.get_state(token)
             return cloudpickle.loads(redis_state)
 
         if token not in self.states:
             self.states[token] = self.state()
         return self.states[token]
 
-    def set_state(self, token: str, state: State):
+    async def set_state(self, token: str, state: State, lock_id: bytes | None = None):
         """Set the state for a token.
 
         Args:
             token: The token to set the state for.
             state: The state to set.
+            lock_id: If provided, the lock_key must be set to this value to set the state.
+
+        Raises:
+            LockExpiredError: If lock_id is provided and the lock for the token is not held by that ID.
+        """
+        if self.redis is None:
+            return
+        if lock_id is not None:
+            lock_key = f"{token}_lock".encode()
```

Comment:
Should we make a function to format the `lock_key` from the token? Since it's used again later. Also maybe this `if` check can be moved to its own function to validate the lock.